File: frontend/src/api/resume.ts
export interface StructuredResumeEntry {
  title: string
  bullets: string[]
}

export interface ParsedResumeData {
  name: string | null
  email: string | null
  phone: string | null
  skills: string[]
  education: string[]
  experience: string[]
  projects: string[]
  leadership: string[]
  experience_entries: StructuredResumeEntry[]
  project_entries: StructuredResumeEntry[]
  leadership_entries: StructuredResumeEntry[]
}

export interface ResumeAnalysis {
  score: number
  summary: string
  strengths: string[]
  warnings: string[]
  improvements: string[]
}

export interface ResumeUploadResponse {
  filename: string
  parsed_data: ParsedResumeData
  analysis: ResumeAnalysis
}

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://127.0.0.1:8000'
const MAX_FILE_SIZE_BYTES = 5 * 1024 * 1024

function validateResumeFile(file: File) {
  const fileName = file.name.toLowerCase()
  const isAllowed = fileName.endsWith('.pdf') || fileName.endsWith('.docx')

  if (!isAllowed) {
    throw new Error('Only PDF and DOCX files are allowed.')
  }

  if (file.size === 0) {
    throw new Error('The selected file is empty.')
  }

  if (file.size > MAX_FILE_SIZE_BYTES) {
    throw new Error('File is too large. Please upload a resume smaller than 5 MB.')
  }
}

export async function uploadResume(file: File): Promise<ResumeUploadResponse> {
  validateResumeFile(file)

  const formData = new FormData()
  formData.append('file', file)

  const response = await fetch(`${API_BASE_URL}/resume/upload`, {
    method: 'POST',
    body: formData,
  })

  if (!response.ok) {
    let errorMessage = 'Failed to upload resume.'

    try {
      const errorData = await response.json()
      errorMessage = errorData.detail || errorMessage
    } catch {
      errorMessage = 'Server error while uploading resume.'
    }

    throw new Error(errorMessage)
  }

  return response.json()
}

File: backend/api/resume.py
from fastapi import APIRouter, File, HTTPException, UploadFile
from services.file_extractor import extract_resume_text
from services.resume_parser import parse_resume_text
from services.resume_scorer import score_resume

router = APIRouter(prefix="/resume", tags=["resume"])

ALLOWED_EXTENSIONS = (".pdf", ".docx")
MAX_FILE_SIZE_BYTES = 5 * 1024 * 1024


@router.post("/upload")
async def upload_resume(file: UploadFile = File(...)):
    filename = file.filename or ""

    if not filename:
        raise HTTPException(status_code=400, detail="No file was uploaded.")

    if not filename.lower().endswith(ALLOWED_EXTENSIONS):
        raise HTTPException(
            status_code=400,
            detail="Invalid file type. Please upload a PDF or DOCX resume.",
        )

    file_bytes = await file.read()

    if not file_bytes:
        raise HTTPException(status_code=400, detail="Uploaded file is empty.")

    if len(file_bytes) > MAX_FILE_SIZE_BYTES:
        raise HTTPException(
            status_code=400,
            detail="File is too large. Please upload a resume smaller than 5 MB.",
        )

    try:
        extracted_text = extract_resume_text(filename, file_bytes)

        if not extracted_text.strip():
            raise HTTPException(
                status_code=400,
                detail="Could not extract text from the uploaded resume.",
            )

        parsed_data = parse_resume_text(extracted_text)
        scored_data = score_resume(parsed_data)

        return {
            "filename": filename,
            "parsed_data": {
                "name": parsed_data["name"],
                "email": parsed_data["email"],
                "phone": parsed_data["phone"],
                "skills": parsed_data["skills"],
                "education": parsed_data["education"],
                "experience": parsed_data["experience"],
                "projects": parsed_data["projects"],
                "leadership": parsed_data["leadership"],
                "experience_entries": parsed_data["experience_entries"],
                "project_entries": parsed_data["project_entries"],
                "leadership_entries": parsed_data["leadership_entries"],
            },
            "analysis": scored_data,
        }

    except HTTPException:
        raise
    except Exception as exc:
        raise HTTPException(
            status_code=500,
            detail=f"Resume processing failed: {str(exc)}",
        )